### DISCUSSION TITLE
Alternative NAIF support through calceph

### DISCUSSION THREAD

**Discussion by mattiaverga** [CelestiaProject/Celestia#1795] [Ideas]
Can [calceph](https://www.imcce.fr/inpop/calceph) be added as alternative to cspice for NAIF support?
I'd like to enable NAIF support for celestia built in Fedora Linux, but there we already have calceph packaged.

**Comment by ajtribick** [comment]
Interesting... I see it's already available in vcpkg so the Windows side of things should hopefully work. It looks like Ubuntu doesn't have it in its repositories so we'd need to build from source, which is already required for cspice.

As far as I'm aware the CeCILL license has compatibility with (some versions of?) the GPL/LGPL so I think the licensing situation there is clearer than it is for cspice. I'd say this is definitely worth a look.

**Comment by 375gnu** [comment]
Pros:
1. Licensed under one of 3 possible free licenses, 2 of which are compatible with GNU GPL. While CSPICE is non-free.
2. Implements not only NAIF, but also INPOPxx and JPL DExxx, so we can remove some code and rely on this library.
3. One C code base for all platforms, not a Fortran code translated into C for a few targets, so it should be more portable.
4. API is cleaner then CSPICE's one.

Cons:
Not found yet.

So for me it sounds like a good idea.

**Comment by ajtribick** [comment]
Looking at this further, I'm not sure if this will work as a full replacement: it does not seem to have support for the kernels describing spacecraft attitude. At least, I am not able to load several of the kernels described in the [Wikibooks SPICE tutorial](https://en.wikibooks.org/wiki/Celestia%2FSPICE) into a test project.

**Comment by mattiaverga** [comment]
Upstream calceph maintainer is usually open and reactive to requests I made when I discover some build problem introduced by dependencies upgrades. If you think it is worth doing, you can try to ask them to add support for those kernels (or check why they don't work).

**Reply by 375gnu** [reply]
@ajtribick have you tried to contact calceph maintainers? If not then I'll do.

**Reply by ajtribick** [reply]
Haven't contacted them, no.

**Reply by 375gnu** [reply]
I have contacted `calceph` maintainer and he replied that *.tf kernels are supported but their implementation has a bug that will be fixed in the next release. Spacecraft clock and leap second kernels will be supported in the future, but without any exact time frame.

But he also shared that their team is working on another library, `OPALE`, which is a `calceph` based C++17 library which will implement the most of `cspice` functions. It will be published in 2024 or 2025.

**Reply by 375gnu** [reply]
just in case https://mantisbt.imcce.fr/ -- that's their buzgilla and https://gitlab.obspm.fr/imcce_calceph/calceph -- their git

**Comment by ajtribick** [comment]
Taking another look at this from the description on the [Wikibook page](https://en.wikibooks.org/wiki/Celestia%2FSPICE), it looks like the main missing pieces are CK and SCLK kernels
